drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,294 @@
+    pub const SPICR: *mut u32 = NOR.wrapping_add(0x0);
+    pub const SPISR: *mut u32 = NOR.wrapping_add(0x1);
+    pub const ADDR: *mut u32 = NOR.wrapping_add(0x2);
+    pub const DUMMY_CYCLES: *mut u32 = NOR.wrapping_add(0x3);
+    pub const DATA_BYTES: *mut u32 = NOR.wrapping_add(0x4);
+    pub const INSTR: *mut u32 = NOR.wrapping_add(0x5);
+    pub const TX_FIFO: *mut u32 = NOR.wrapping_add(0x6);
+    pub const RX_FIFO: *mut u32 = NOR.wrapping_add(0x7);
+}
+
+#[allow(unused)]
+mod instr {
+    pub const PAGE_PROGRAM: u32 = 0x02;
+    pub const READ: u32 = 0x03;
+    pub const READ_STATUS_1: u32 = 0x05;
+    pub const READ_STATUS_2: u32 = 0x35;
+    pub const READ_STATUS_3: u32 = 0x15;
+    pub const WRITE_STATUS_2: u32 = 0x31;
+    pub const WRITE_ENABLE: u32 = 0x06;
+    pub const FAST_READ_QUAD: u32 = 0x6b;
+    pub const FAST_READ_QUAD_OUTPUT_4B: u32 = 0x6c;
+    pub const SECTOR_ERASE: u32 = 0x20;
+    pub const READ_JEDEC_ID: u32 = 0x9f;
+    pub const BLOCK_ERASE_64KB: u32 = 0xd8;
+    pub const BLOCK_ERASE_64KB_4B: u32 = 0xdc;
+    pub const QUAD_INPUT_PAGE_PROGRAM: u32 = 0x32;
+    pub const QUAD_INPUT_PAGE_PROGRAM_4B: u32 = 0x34;
+}
+
+impl ServerImpl {
+    fn flash_read_id(&self) -> [u8; 20] {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 20);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_JEDEC_ID);
+        self.wait_fpga_busy();
+        let mut out = [0u8; 20];
+        for i in 0..out.len() / 4 {
+            let v = self.read_reg(reg::RX_FIFO);
+            for (j, byte) in v.to_le_bytes().iter().enumerate() {
+                out[i * 4 + j] = *byte;
+            }
+        }
+        out
+    }
+
+    fn read_reg(&self, reg: *mut u32) -> u32 {
+        unsafe { reg.read_volatile() }
+    }
+
+    fn write_reg(&self, reg: *mut u32, v: u32) {
+        unsafe { reg.write_volatile(v) };
+    }
+
+    /// Wait until the FPGA is idle
+    fn wait_fpga_busy(&self) {
+        loop {
+            let status = self.read_reg(reg::SPISR);
+            if (status & 1) == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy(status));
+            sleep_for(1);
+        }
+    }
+
+    /// Wait until a word is available in the RX buffer
+    fn wait_fpga_rx(&self) {
+        for i in 0.. {
+            let status = self.read_reg(reg::SPISR);
+            // Wait for the
+            const RX_EMPTY_BIT: u32 = 1 << 6;
+            if status & RX_EMPTY_BIT == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy(status));
+            // Initial busy-loop for faster response
+            if i >= 32 {
+                sleep_for(1);
+            }
+        }
+    }
+
+    /// Wait until the SPI flash is idle
+    fn wait_flash_busy(&self, t: Trace) {
+        // Wait for the busy flag to be unset
+        while (self.read_flash_status() & 1) != 0 {
```

Comment:
**nitpick:** Other loops do `let status = act(); if act & compare == 0 { break; }`, but this uses a `while (act & compare != 0`. Oh the horror! :)

Reply:
There's a reason for the difference: the other loops actually log the status in the ringbuf, so we need it in a variable.

Reply:
Oh, indeed. My bad.